### PR TITLE
[Nodes UI] More work

### PR DESCRIPTION
**Pull request by blessedcoolant**
- Add `CodeFormer` to the Face Restore Node.
- Fix a small bug with the NodeSearch keys not interacting correctly with tag results.
- Reenabled the `Load Image` node.
- Add `Float Value`, `String Value`, `Boolean Value` parameter nodes.
- Changed `params` to `values`.
- Added `Negative Prompt` to all Template / Graph nodes.

### CODE REVIEW

File: invokeai/app/invocations/params.py [hunk 1]
```diff
@@ -2,17 +2,131 @@
...
+    class Config(InvocationConfig):
+        schema_extra = {
+            "ui": {
+                "tags": ["parameters", "merge", "string"],
+                "title": "Merge String"
+            }
+        }
+
+    def invoke(self, context: InvocationContext) -> StringOutput:
+        return StringOutput(a=f'{self.a} {self.b}')
```
Comment (psychedelicious) [inline invokeai/app/invocations/params.py:132]: I don't think this should be a node - we should design a more generalized way to concatenate strings, ideally using multiple iterate/collect nodes (or similar), which let you do distributed concatenation over multiple arrays. 

The simplest use case of such a node would be concatenating two strings, superseding this node, but it could also do much more.

**Review by psychedelicious — CHANGES_REQUESTED** [review]
1 requested change - but otherwise looks fine to me, just with the caveat that we will want to refactor how the restoration nodes work entirely at some point.

File: invokeai/app/invocations/params.py (hunk 1 above)
Comment (blessedcoolant) [reply invokeai/app/invocations/params.py:132]: Okie removing it.

**Review by psychedelicious — APPROVED** [review]

**Comment by blessedcoolant** [comment]
Needs @Kyle0654

File: invokeai/app/invocations/image.py [hunk 2]
```diff
@@ -80,18 +80,19 @@ class LoadImageInvocation(BaseInvocation):
 
     # fmt: off
     type: Literal["load_image"] = "load_image"
-
     # Inputs
-    image_type: ImageType = Field(description="The type of the image")
-    image_name:       str = Field(description="The name of the image")
+    image: ImageField = Field(description="The type of the image")
```
Comment (Kyle0654) [inline invokeai/app/invocations/image.py:84]: This may make it difficult to use in the CLI. Note that this node is effectively just "create an image field".

File: invokeai/app/services/restoration_services.py [hunk 3]
```diff
@@ -62,13 +64,16 @@ def upscale_and_reconstruct(
                                     ">> GFPGAN not found. Face restoration is disabled."
                                 )
                             else:
-                                image = self.gfpgan.process(image, strength, seed)
+                                print(">> Restoring with: GFPGAN")
```
Comment (Kyle0654) [inline invokeai/app/services/restoration_services.py:67]: Didn't @lstein set up logging recently? We should be avoiding print in favor of logging, if so.

File: invokeai/app/services/restoration_services.py [hunk 4]
```diff
@@ -62,13 +64,16 @@ def upscale_and_reconstruct(
...
+                                image = self.gfpgan.process(
+                                    image, strength, seed)
                         if facetool == "codeformer":
                             if self.codeformer is None:
                                 print(
                                     ">> CodeFormer not found. Face restoration is disabled."
                                 )
                             else:
+                                print(">> Restoring with: CodeFormer")
```
Comment (Kyle0654) [inline invokeai/app/services/restoration_services.py:76]: Same comment about logging (and the line above probably needs updating).

File: invokeai/frontend/web/src/features/nodes/util/parseSchema.ts [hunk 5]
```diff
@@ -13,7 +13,7 @@ import {
   buildOutputFieldTemplates,
 } from './fieldTemplateBuilders';
 
-const invocationBlacklist = ['Graph', 'Collect', 'LoadImage'];
+const invocationBlacklist = ['Graph', 'Collect'];
```
Comment (Kyle0654) [inline invokeai/frontend/web/src/features/nodes/util/parseSchema.ts:16]: Prefer "DenyList" for inclusivity.

**Review by Kyle0654 — CHANGES_REQUESTED** [review]

File: invokeai/app/services/restoration_services.py (hunk 3 above)
Comment (blessedcoolant) [reply invokeai/app/services/restoration_services.py:67]: That is not merged yet. Can port it over then.

File: invokeai/frontend/web/src/features/nodes/util/parseSchema.ts (hunk 5 above)
Comment (blessedcoolant) [reply invokeai/frontend/web/src/features/nodes/util/parseSchema.ts:16]: This will need more changes than just in this file. I'll tag @psychedelicious so he can fully replace them across the board.

File: invokeai/app/invocations/image.py (hunk 2 above)
Comment (psychedelicious) [reply invokeai/app/invocations/image.py:84]: It's useful in the UI when you want to pass the same image to multiple image inputs

File: invokeai/frontend/web/src/features/nodes/util/parseSchema.ts (hunk 5 above)
Comment (psychedelicious) [reply invokeai/frontend/web/src/features/nodes/util/parseSchema.ts:16]: Oof, good call

File: invokeai/app/invocations/image.py (hunk 2 above)
Comment (Kyle0654) [reply invokeai/app/invocations/image.py:84]: You can't construct this data structure in the CLI though.

Comment (psychedelicious) [reply invokeai/app/invocations/image.py:84]: Hmm. So there's no way to do img2img on the cli with a user-provided image?

Comment (StAlKeR7779) [reply invokeai/app/invocations/image.py:84]: Maybe `Image loading` node fits here?
This node will load image and add to cache with this name, and it's output is `ImageField`.

Comment (StAlKeR7779) [reply invokeai/app/invocations/image.py:84]: > It's useful in the UI when you want to pass the same image to multiple image inputs

If I understand correctly - if we have `ImageField` we already can pass it to multiple nodes, isn't it?
image

And to load new image to graph we using this node with simple values without `ImageField`.

Comment (Kyle0654) [reply invokeai/app/invocations/image.py:84]: Yah load image is just to take a type and name and create an `ImageField`, which can then be passed to other things. Basically how you'd load an uploaded file into the pipeline.

Comment (psychedelicious) [reply invokeai/app/invocations/image.py:84]: Even if LoadImage has a type and name as inputs instead of an ImageField, as you've suggested, it still doesn't let users on the cli take an image from anywhere on disk and use it. That use case is not addressed by LoadImage regardless.

That aside, it's much simpler for LoadImage to have an ImageField than a name and type.

Then the UI can present it like any other image field input (which allows you to drag an image from the gallery into it, or in the future, drag an image from the file system onto it to upload).

If it's just a type and name, the ui needs to have special handling for this specific node.

Comment (Kyle0654) [reply invokeai/app/invocations/image.py:84]: *goes to look at load_image node*...

```py
# TODO: this isn't really necessary anymore
```

Well then... still concerned about how to do this with the UI, but maybe this node should just be deleted (and create a separate "param" node for images, like the int/float/string params, since that's what you seem to intend to use this like).

Comment (psychedelicious) [reply invokeai/app/invocations/image.py:84]: Yeah that's how it's being used. @blessedcoolant lets follow that suggestion and make this a "param" node. 

We will have to figure out a CLI solution for loading an image from disk. May need to just have some nodes that are cli-only or ui-only

Comment (Kyle0654) [reply invokeai/app/invocations/image.py]: That's what commands are for in the CLI. Already have a bunch of different commands supported =)

File: invokeai/app/invocations/values.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+class IntOutput(BaseInvocationOutput):
+    """An integer output"""
+    #fmt: off
+    type: Literal["int_output"] = "int_output"
+    a: int = Field(default=None, description="The output integer")
+    #fmt: on
+
+
+class IntegerValueInvocation(BaseInvocation):
+    """An integer parameter"""
```
Comment (psychedelicious) [inline invokeai/app/invocations/values.py:20]: Let's fix all the docstrings to say `value`

File: invokeai/app/invocations/values.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class ImageValueInvocation(BaseInvocation):
+    """Load an image and provide it as output."""
+
+    # fmt: off
+    type: Literal["value_image"] = "value_image"
+    # Inputs
+    image: ImageField = Field(description="The input image")
+    # fmt: on
```
Comment (psychedelicious) [inline invokeai/app/invocations/values.py:126]: Also the docstring and description here should be updated

File: invokeai/frontend/web/src/features/nodes/components/search/NodeSearch.tsx
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/nodes/components/search/NodeSearch.tsx:1]: This needs to be within the `TopLeftPanel` component. `reactflow` expects a single panel at each position, we should have a single component for each position, named after that position, even if it only has a single component inside it.

**Review by psychedelicious — CHANGES_REQUESTED** [review]
Requested a couple small changes.

Also the test fixtures need to be updated for negative prompts

File: invokeai/app/invocations/values.py (hunk 7 above)
Comment (blessedcoolant) [reply invokeai/app/invocations/values.py:126]: What should these be updated to?

**Comment by hipsterusername** [comment]
@blessedcoolant - Looks like some conflicts to resolve here, then hopefully @Kyle0654 can take a look and merge
